### PR TITLE
[ads] General optimizations

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/34897

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_ads/core/internal/account/confirmations/payload/reward_confirmation_payload_util.cc [hunk 1]
```diff
@@ -22,6 +22,8 @@ constexpr char kPublicKeyKey[] = "publicKey";
 
 base::Value::Dict BuildRewardConfirmationPayload(const RewardInfo& reward) {
   base::Value::List list;
+  list.reserve(1);
```
Comment (aseren) [inline components/brave_ads/core/internal/account/confirmations/payload/reward_confirmation_payload_util.cc:25]: no need to reserve place for 1 element: there is no optimization for such a case

File: components/brave_ads/core/internal/account/issuers/issuers_value_util.cc [hunk 2]
```diff
@@ -156,7 +156,7 @@ std::optional<IssuerList> ValueToIssuers(const base::Value::List& list) {
     issuer.type = *issuer_type;
     issuer.public_keys = *public_keys;
 
-    issuers.push_back(issuer);
+    issuers.emplace_back(issuer);
```
Comment (aseren) [inline components/brave_ads/core/internal/account/issuers/issuers_value_util.cc:159]:
```
So in general, if both push_back() and emplace_back() would work with the same arguments, you should prefer push_back(), and likewise for insert() vs. emplace().
```
https://abseil.io/tips/112

Comment (tmancey) [reply components/brave_ads/core/internal/account/issuers/issuers_value_util.cc:160]: Resolved :-)

File: components/brave_ads/core/internal/common/time/time_constraint_util.cc [hunk 3]
```diff
@@ -14,12 +13,31 @@ bool DoesHistoryRespectRollingTimeConstraint(
...
-        return base::Time::Now() - created_at < time_constraint;
-      });
+  const base::Time now = base::Time::Now();
 
-  return count < cap;
+  size_t count = 0;
+  size_t remaining_cap = cap;
+
+  for (auto iter = history.crbegin(); iter != history.crend();
+       ++iter, --remaining_cap) {
```
Comment (aseren) [inline components/brave_ads/core/internal/common/time/time_constraint_util.cc:22]: I didn't understand how the `base::ranges::count_if` was converted to this loop.
can `history` size be higher then `remaining_cap` value? If so, then we can end up with decrementing `remaining_cap` with zero value.

File: components/brave_ads/core/internal/common/url/url_request_string_util.cc [hunk 4]
```diff
@@ -30,46 +29,42 @@ bool ShouldAllowHeader(const std::string& header) {
 
 std::string HeadersToString(const std::vector<std::string>& headers,
                             const int indent = 4) {
-  std::vector<std::string> formatted_headers;
+  std::ostringstream formatted_headers;
```
Comment (aseren) [inline components/brave_ads/core/internal/common/url/url_request_string_util.cc:32]: I'm not sure that streams should be used for such things. `the libraries in absl/strings or the equivalent are usually a better choice than std::stringstream.`

https://google.github.io/styleguide/cppguide.html#Streams

File: components/brave_ads/core/internal/common/time/time_constraint_util.cc (hunk 3 above)
Comment (tmancey) [reply components/brave_ads/core/internal/common/time/time_constraint_util.cc]: Optimize again :-) Please review. Thanks

File: components/brave_ads/core/internal/common/url/url_request_string_util.cc (hunk 4 above)
Comment (tmancey) [reply components/brave_ads/core/internal/common/url/url_request_string_util.cc]: Reverted the stream changes, thanks

File: components/brave_ads/core/internal/common/time/time_constraint_util.cc [hunk 5]
```diff
@@ -5,21 +5,29 @@
...
+  const base::Time threshold = base::Time::Now() - time_constraint;
 
-  return count < cap;
+  for (auto iter = history.crbegin(); iter != history.crend(); ++iter) {
+    if (*iter <= threshold) {
+      // If the time point is older than the threshold, break the loop.
+      break;
+    }
+
+    --cap;
```
Comment (aseren) [inline components/brave_ads/core/internal/common/time/time_constraint_util.cc:24]: this can lead to insigned underflow if cap value is already 0

Comment (tmancey) [reply components/brave_ads/core/internal/common/time/time_constraint_util.cc:29]: Great catch, thank you

File: components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/daypart_exclusion_rule_util.cc [hunk 6]
```diff
@@ -13,7 +14,8 @@ namespace brave_ads {
 
 bool MatchDayOfWeek(const CreativeDaypartInfo& daypart,
                     const char day_of_week) {
-  return daypart.days_of_week.find(day_of_week) != std::string::npos;
+  return std::find(daypart.days_of_week.cbegin(), daypart.days_of_week.cend(),
```
Comment (aseren) [inline components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/daypart_exclusion_rule_util.cc:17]: can we use base::Contains here?

Comment (tmancey) [reply components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/daypart_exclusion_rule_util.cc]: Changed, thanks

File: components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/exclusion_rules_base.h [hunk 7]
```diff
@@ -42,7 +42,7 @@ class ExclusionRulesBase {
   std::vector<std::unique_ptr<ExclusionRuleInterface<CreativeAdInfo>>>
       exclusion_rules_;
 
-  std::set<std::string> uuids_;
+  std::unordered_set<std::string> uuids_;
```
Comment (aseren) [inline components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/exclusion_rules_base.h:45]: is it better to use std::set here?

File: components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/exclusion_rules_util.h [hunk 8]
```diff
@@ -17,35 +15,34 @@
...
-
-  return filtered_creative_ads;
+  if (creative_ads.empty()) {
+    return;
+  }
+
+  const bool can_cap_last_served_creative_ad =
+      CanCapLastServedCreativeAd(creative_ads);
+
+  creative_ads.erase(
```
Comment (aseren) [inline components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/exclusion_rules_util.h:35]: base::EraseIf

File: components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/exclusion_rules_base.h (hunk 7 above)
Comment (tmancey) [reply components/brave_ads/core/internal/serving/eligible_ads/exclusion_rules/exclusion_rules_base.h:45]: For real-world data we found that flat_set was approx. 8% faster. However, it has been decided to followup across ads to investigate and optimize where needed.

**Pull request merged**